lenskit-eval/src/main/java/org/lenskit/eval/traintest/recommend/TopNNDPMMetric.java

```diff
@@ -0,0 +1,139 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.eval.traintest.recommend;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import it.unimi.dsi.fastutil.longs.*;
+import org.apache.commons.lang3.StringUtils;
+import org.grouplens.lenskit.util.statistics.MeanAccumulator;
+import org.lenskit.api.ResultList;
+import org.lenskit.eval.traintest.AlgorithmInstance;
+import org.lenskit.eval.traintest.DataSet;
+import org.lenskit.eval.traintest.TestUser;
+import org.lenskit.eval.traintest.metrics.MetricResult;
+import org.lenskit.specs.AbstractSpec;
+import org.lenskit.util.collections.LongUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+
+/**
+ * Measure the nDPM of the top-N recommendations, using rankings.
+ * This metric is registered with the type name `ndpm`.
+ * The paper used as a reference for this implementation is http://www2.cs.uregina.ca/~yyao/PAPERS/jasis_ndpm.pdf.
+ * Created by VaibhavMahant on 2/1/16.
+ */
+public class TopNNDPMMetric extends TopNMetric<MeanAccumulator> {
+    public static final String DEFAULT_COLUMN = "TopN.nDPM";
+
+    /**
+     * Construct a top-N nDCG metric from a spec.
+     * @param spec The spec.
+     */
+    @JsonCreator
+    public TopNNDPMMetric(Spec spec) {this(spec.getColumnName());
+    }
+
+    /**
+     * Construct a new nDPM Top-N metric.
+     */
+    public TopNNDPMMetric(String name) {
+        super(Collections.singletonList(StringUtils.defaultString(name, DEFAULT_COLUMN)),
+              Collections.singletonList(StringUtils.defaultString(name, DEFAULT_COLUMN)));
+    }
+
+    @Nullable
+    @Override
+    public MeanAccumulator createContext(AlgorithmInstance algorithm, DataSet dataSet, org.lenskit.api.Recommender recommender) {
+        return new MeanAccumulator();
+    }
+
+    @Nonnull
+    @Override
+    public MetricResult getAggregateMeasurements(MeanAccumulator context) {
+        return MetricResult.singleton(DEFAULT_COLUMN, context.getMean());
+    }
+
+    @Nonnull
+    @Override
+    public MetricResult measureUser(TestUser user, int targetLength, ResultList recommendations, MeanAccumulator context) {
+        if (recommendations == null) {
+            return MetricResult.empty();
+        }
+
+        Long2DoubleMap ratings = user.getTestRatings();
+
+        long[] actual = LongUtils.asLongCollection(recommendations.idList()).toLongArray();
+
+        double dpm = computeDPM(actual, ratings);
+
+        double nDPM = dpm / 2 * dpm; // Normalized nDPM
```

Comment:
Is this correct? My read of the paper is that the denominator should be 2C. Also, when dividing by a product, we must put the denominator in parentheses.